src/matching.rs

```diff
@@ -113,16 +118,32 @@ impl<T, Impl: SelectorImpl> SelectorMap<T, Impl> {
 
         SelectorMap::get_matching_rules(element,
                                         parent_bf,
-                                        &self.universal_rules,
+                                        &self.other_rules,
                                         matching_rules_list,
                                         shareable);
 
         // Sort only the rules we just added.
         sort_by(&mut matching_rules_list[init_len..], &compare);
+    }
 
-        fn compare<T>(a: &DeclarationBlock<T>, b: &DeclarationBlock<T>) -> Ordering {
-            (a.specificity, a.source_order).cmp(&(b.specificity, b.source_order))
+    /// Append to `rule_list` all universal Rules (rules with selector `*|*`) in
+    /// `self` sorted by specifity and source order.
+    pub fn get_universal_rules<V>(&self,
+                                  matching_rules_list: &mut V)
+                                  where V: VecLike<DeclarationBlock<T>> {
+        if self.empty {
+            return
         }
+
+        let init_len = matching_rules_list.len();
+
+        for rule in self.other_rules.iter() {
+            if rule.selector.simple_selectors.is_empty() {
```

Comment:
Did you mean to include rules with selectors like `#foo > *|*`? If so, please change the doc-comment from "rules with selectors *|*" to "rules whose left-most compound selector (separated by combinators) is *|*". If not, this needs to check `rule.selector.next.is_none()` as well.